backend/file-system/src/main/java/com/htec/filesystem/service/FileService.java

```diff
@@ -204,45 +205,89 @@ public void updateDeletedFiles(AuthUser user, List<Long> fileIds, Boolean delete
 
     void recover(AuthUser user, List<FileEntity> fileEntities) {
 
+        for (FileEntity fileEntity : fileEntities) {
+            Long parentFolderId = fileEntity.getParentFolderId();
+            Optional<FolderEntity> folderEntity = folderRepository.findById(parentFolderId);
+            if (!folderEntity.isPresent())
+                fileEntity.setParentFolderId(null);
+            if (folderEntity.get().getDeleted())
+                fileEntity.setParentFolderId(null);
+        }
+
         List<Long> folderIds = fileEntities.stream().map(FileEntity::getParentFolderId).collect(Collectors.toList());
 
         List<FileEntity> fileEntitiesNotDeleted = fileRepository.findAllByUserIdAndNotDeletedAndParentFolderIdsIn(user.getId(), folderIds);
 
-        Map<Optional<Long>, List<FileEntity>> filesByFolder = fileEntitiesNotDeleted.stream().collect(Collectors.groupingBy(e -> Optional.ofNullable(e.getParentFolderId())));
+        Map<Optional<Long>, List<FileEntity>> filesByFolder = fileEntitiesNotDeleted.stream().
```

Comment:
you can just filter null values for parentFolderId and than you can avoid having Optional as a key in a map